src/linear_algebra/decomposition.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2021 Hanting Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Hanting Zhang
+-/
+import algebra.direct_sum.module
+
+/-!
+# Decompositions of Modules
+This file talks about decompositions of modules.
+-/
+
+universes u v w
+
+open_locale direct_sum classical big_operators
+open classical linear_map dfinsupp direct_sum
+
+namespace submodule
+
+/--
+A `decomposition ι R M` for a module `M` is the type of `ι`-indexed decompositions of `M`.
+The factors are available as `factors : ι → submodule R M`.
+To create a decomposition of `M`, one furthermore must provide data that the
+factors satisfy `∀ i : ι, disjoint (factors i) (⨆ (j ≠ i), factors j)` and
+that their union is the whole module `M`.
+
+Note: Why not say `set (submodule R M)`? Then one could get rid of explicitly
+identifying `ι`. But I think it's better to follow the `basis` design,
+which phased out of the `set` design because of problems with it.
+-/
+@[nolint has_inhabited_instance]
+structure decomposition (ι : Type u) (R : Type v) (M : Type w)
```

Comment:
This bears a resemblance to #10115. Both approaches have their merits, but I hope we can get the best of both.